rstudio/c9s-python-3.11/setup_rstudio.py

```diff
@@ -46,6 +46,30 @@ def _support_arg(arg):
     ret = subprocess.check_output([get_rstudio_executable('rserver'), '--help'])
     return ret.decode().find(arg) != -1
 
+def detect_env():
+    import socket
+    supports_ipv4 = supports_ipv6 = False
+    try:
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            s.bind(('127.0.0.1', 0))
+            supports_ipv4 = True
+    except OSError:
+        pass
+    try:
+        with socket.socket(socket.AF_INET6, socket.SOCK_STREAM) as s:
+            s.bind(('::1', 0))
+            supports_ipv6 = True
+    except OSError:
+        pass
+    if supports_ipv4 and supports_ipv6:
+        return '::'  # Dual-stack
+    elif supports_ipv6:
+        return '::'
+    elif supports_ipv4:
+        return '0.0.0.0'
```

Comment:
I'm inclined to replace this with `return '::'`, because that _should_ work just everywhere; We have ubi9 images that do support ipv6; and jupyterlab/etc. all use `socket.AF_INET6` in effect. But your version feels somewhat more cautious and safer.